### DISCUSSION TITLE
How can I fetch an access token only when I get a 401 response?

### DISCUSSION THREAD

**Discussion by nerdroid23** [saloonphp/saloon#77] [Q&A] [answered]
#### I have two connectors, the first one is to fetch an access token and the second one is to interact with the resources.

<details>
<summary>app\Http\Integrations\AutoConnect\AutoConnectAccessTokenConnector.php</summary>

```php
<?php

namespace App\Http\Integrations\AutoConnect;

use Sammyjo20\Saloon\Http\SaloonConnector;
use Sammyjo20\Saloon\Traits\Plugins\AcceptsJson;

class AutoConnectAccessTokenConnector extends SaloonConnector
{
    use AcceptsJson;

    public function defineBaseUrl(): string
    {
        return config('services.autoconnect.oauthUrl');
    }
}
```
</details>

<details>
<summary>app\Http\Integrations\AutoConnect\AutoConnectConnector.php</summary>

```php
<?php

namespace App\Http\Integrations\AutoConnect;

use App\Http\Integrations\AutoConnect\Requests\GetAccessTokenRequest;
use Sammyjo20\Saloon\Http\SaloonConnector;
use Sammyjo20\Saloon\Http\SaloonRequest;
use Sammyjo20\Saloon\Http\SaloonResponse;
use Sammyjo20\Saloon\Traits\Plugins\AcceptsJson;

class AutoConnectConnector extends SaloonConnector
{
    use AcceptsJson;

    // public function boot(SaloonRequest $request): void
    // {
    //     $this->addResponseInterceptor(function (SaloonRequest $request, SaloonResponse $response) {
    //         if ($response->failed()) {
    //             // $authToken = GetAccessTokenRequest::make()->send()->body();
    //             $request->withTokenAuth(
    //                 GetAccessTokenRequest::make()->send()->body()
    //             )->send();
    //         }
    //     });
    // }

    public function defineBaseUrl(): string
    {
        return config('services.autoconnect.baseUrl');
    }
}
```
</details>

#### Then I have my requests, the first one fetches the access token and the second one a list of states from the api.

<details>
<summary>app\Http\Integrations\AutoConnect\Requests\GetAccessTokenRequest.php</summary>

```php
<?php

namespace App\Http\Integrations\AutoConnect\Requests;

use App\Http\Integrations\AutoConnect\AutoConnectAccessTokenConnector;
use Sammyjo20\Saloon\Constants\Saloon;
use Sammyjo20\Saloon\Http\SaloonRequest;
use Sammyjo20\Saloon\Traits\Plugins\HasJsonBody;

class GetAccessTokenRequest extends SaloonRequest
{
    use HasJsonBody;

    protected ?string $connector = AutoConnectAccessTokenConnector::class;

    protected ?string $method = Saloon::POST;

    public function defineEndpoint(): string
    {
        return '/oauth2/token';
    }

    public function defaultData(): array
    {
        return [
            'client_id' => config('services.autoconnect.clientId'),
            'client_secret' => config('services.autoconnect.clientSecret'),
            'grant_type' => 'client_credential',
            'scope' => 'SEMI',
        ];
    }
}
```
</details>

<details>
<summary>app\Http\Integrations\AutoConnect\Requests\ListStatesRequest.php</summary>

```php
<?php

namespace App\Http\Integrations\AutoConnect\Requests;

use App\Http\Integrations\AutoConnect\AutoConnectConnector;
use Sammyjo20\Saloon\Constants\Saloon;
use Sammyjo20\Saloon\Http\SaloonRequest;

class ListStatesRequest extends SaloonRequest
{
    protected ?string $connector = AutoConnectConnector::class;

    protected ?string $method = Saloon::GET;

    public function defineEndpoint(): string
    {
        return '/Catalogos/Ubicacion/Estados';
    }
}
```
</details>

#### Checking in the docs and I found that adding [`addResponseInterceptor()`](https://docs.saloon.dev/advanced/response-interceptors) in `boot()` of `AutoConnectConnector`, so I added the following
```php
public function boot(SaloonRequest $request): void
{
    $this->addResponseInterceptor(function (SaloonRequest $request, SaloonResponse $response) {
        if ($response->failed()) {
            $authToken = GetAccessTokenRequest::make()->send()->body();
            $request->withTokenAuth($authToken)->send();
        }
    });
}
```

### Thanks!

**Comment by Sammyjo20** [comment]
Hey!

How does the token come back from the first request? Is the body JSON or is it just raw plain text that comes back from the API?

**Reply by nerdroid23** [reply]
Thanks for the quick response!

It comes back as plain raw text

**Reply by Sammyjo20** [reply]
I think what you're after is actually to place the logic where you fetch a token in the `boot` method and not inside of a `responseInterceptor` inside a boot method.

The boot method runs before the request is sent, the response interceptor runs after the response has come back. Try that! :D

**Reply by nerdroid23** [reply]
I think it is what I want? Since I only want to fetch the access token when I get a 401. The boot method would not give me access to the response, would it? The response interceptor let's me check the status of the response and based on that I can either fetch an access token and then resend the request with the freshly fetched token.

**Comment by nerdroid23** [comment, marked as answer]
Figured it out! Instead of `GetAccessTokenRequest::make()->send()->body()`, I had to call `GetAccessTokenRequest::make()->send()->json()`. `body()` was returning the token in wrapped in a excess pair of double quotes `""super-secure-token-goes-here""` and `json()` just returns the plain string `"super-secure-token-goes-here"`.
